docs/references/api/resource_embedding.rst

```diff
@@ -1189,6 +1199,39 @@ You can use this to get the columns of a join table in a many-to-many relationsh
      }
    ]
 
+One-to-many and Many-to-many relationships
+------------------------------------------
+
+For these relationships, the spread embedded resource columns will be aggregated into arrays.
+
 .. note::
+  This is disabled by default, since aggregates can cause performance issues.
+  It can be activated by setting :ref:`db-aggregates-enabled` to ``true``.
```

Comment:
It's confusing to require this config since the use of an aggregate is an implementation detail. The config only makes sense if users want to explicitly use an agg function. Is the impact on performance that high?

Reply:
Under the hood, the feature does a `GROUP BY` for any column that is selected, then does an aggregate `array_agg()` to every spread column. So I don't think it's that different from doing a `count()` aggregate with `GROUP BY`, for instance. Although I tried to explain that doing so uses an "aggregate into arrays", I get that it can be confusing, maybe mention that `array_agg()` with `GROUP BY` is used? or perhaps it's better to add another configuration?

Reply:
Comparing the plan of a simple embedding:

```
$ curl "localhost:3000/clients?select=name,projects(id,name)" -H "Accept: application/vnd.pgrst.plan"

Aggregate  (cost=31833.03..31833.05 rows=1 width=144)
  ->  Nested Loop Left Join  (cost=25.02..31826.68 rows=1270 width=64)
        ->  Seq Scan on clients  (cost=0.00..22.70 rows=1270 width=36)
        ->  Aggregate  (cost=25.02..25.03 rows=1 width=32)
              ->  Seq Scan on projects projects_1  (cost=0.00..25.00 rows=6 width=36)
                    Filter: (client_id = clients.id)
```

Vs this new spread:

```
$ curl "localhost:3000/clients?select=name,...projects(id,name)" -H "Accept: application/vnd.pgrst.plan"

Aggregate  (cost=78.26..78.28 rows=1 width=144)
  ->  HashAggregate  (cost=73.26..77.26 rows=200 width=96)
        Group Key: clients.name
        ->  Hash Right Join  (cost=38.58..63.74 rows=1270 width=68)
              Hash Cond: (projects_1.client_id = clients.id)
              ->  Seq Scan on projects projects_1  (cost=0.00..22.00 rows=1200 width=40)
              ->  Hash  (cost=22.70..22.70 rows=1270 width=36)
                    ->  Seq Scan on clients  (cost=0.00..22.70 rows=1270 width=36)
```

The new spread will be cheaper in theory. So I'm leaning towards no configuration and just enable it.

Reply:
You might want to do some further tests to corroborate this, perhaps by adding 1M rows for the tables, using the pgbench setup https://github.com/PostgREST/postgrest/tree/main/test/pgbench. But so far looks fine to me.

Reply:
I added this to the `data.sql`:

```sql
insert into test.clients(id,name) select x.a, 'Client ' || x.a from generate_series(3,30000) x(a);
insert into test.projects(id,name,client_id) select x.a, 'Project ' || x.a, x.a/2 from generate_series(6,60000) x(a);
insert into test.users(id,name) select x.a, 'User ' || x.a from generate_series(4,30000) x(a);
insert into test.users_projects(user_id,project_id) select x.a, x.a/2 from generate_series(4,30000) x(a);
```

1. It's much faster and more performant when the to-many spread is used in the top resource. For 30k `clients` and 60k `projects`:

    <details><summary>Plan</summary>
    <p>
    
    ```console
    $ # to-many spread embed
    $ curl "localhost:3000/clients?select=*,...projects(pi:id,pn:name)" -H "Accept: application/vnd.pgrst.plan"
    
    Aggregate  (cost=8109.33..8109.35 rows=1 width=144)
      ->  GroupAggregate  (cost=6759.32..7959.32 rows=30000 width=80)
            Group Key: clients.id
            ->  Sort  (cost=6759.32..6909.32 rows=60000 width=33)
                  Sort Key: clients.id
                  ->  Hash Right Join  (cost=857.00..1997.52 rows=60000 width=33)
                        Hash Cond: (projects_1.client_id = clients.id)
                      ->  Seq Scan on projects projects_1  (cost=0.00..983.00   rows=60000 width=21)
                        ->  Hash  (cost=482.00..482.00 rows=30000 width=16)
                            ->  Seq Scan on clients  (cost=0.00..482.00   rows=30000 width=16)
  
    $ # normal embed
    $ curl "localhost:3000/clients?select=*,projects(pi:id,pn:name)" -H "Accept: application/vnd.pgrst.plan"
  
    Aggregate  (cost=33991607.00..33991607.02 rows=1 width=144)
      ->  Nested Loop Left Join  (cost=1133.01..33991457.00 rows=30000 width=48)
            ->  Seq Scan on clients  (cost=0.00..482.00 rows=30000 width=16)
            ->  Aggregate  (cost=1133.01..1133.02 rows=1 width=32)
                ->  Seq Scan on projects projects_1  (cost=0.00..1133.00 rows=2   width=17)
                        Filter: (client_id = clients.id)
    ```
    
    </p>
    </details> 
  
    <details><summary>Responses</summary>
    <p>
    
    ```bash
    $ # to-many spread embed
    $ curl "localhost:3000/clients?select=*,...projects(pi:id,pn:name)" -i
    
    HTTP/1.1 200 OK
    Transfer-Encoding: chunked
    Date: Wed, 09 Oct 2024 22:28:37 GMT
    Server: postgrest/12.3 (pre-release)
    Content-Range: 0-29999/*
    Content-Location: /clients?select=%2A%2Cprojects%28pi%3Aid%2Cpn%3Aname%29
    Content-Type: application/json; charset=utf-8
    Server-Timing: jwt;dur=0.0, parse;dur=0.3, plan;dur=0.2, transaction;dur=238.6, response;dur=0.0
  
    ...
  
    $ # normal embed
    $ curl "localhost:3000/clients?select=*,projects(pi:id,pn:name)" -i
  
    HTTP/1.1 200 OK
    Transfer-Encoding: chunked
    Date: Wed, 09 Oct 2024 22:28:51 GMT
    Server: postgrest/12.3 (pre-release)
    Content-Range: 0-29999/*
    Content-Location: /clients?select=%2A%2C...projects%28pi%3Aid%2Cpn%3Aname%29
    Content-Type: application/json; charset=utf-8
    Server-Timing: jwt;dur=0.0, parse;dur=0.3, plan;dur=0.2, transaction;dur=5.1, response;dur=0.0
    
    ...
  
    ```
    
    </p>
    </details> 
  
    The to-many spread took some milliseconds to respond. The normal embed took 1:30~ minutes.

2. When the to-many spread is nested, the cost is higher (1.5x) than the normal embed, but it takes roughly the same time to complete. I added 30k `users` and `users_projects`.

    <details><summary>Plan</summary>
    <p>
    
    ```console
    $ # to-many spread embed
    $ curl "localhost:3000/clients?select=*,projects(pi:id,pn:name,...users(ui:id,un:name))" -H "Accept: application/vnd.pgrst.plan"
  
    Aggregate  (cost=52294208.81..52294208.83 rows=1 width=144)
      ->  Nested Loop Left Join  (cost=1743.09..52294058.81 rows=30000 width=48)
            ->  Seq Scan on clients  (cost=0.00..482.00 rows=30000 width=16)
            ->  Aggregate  (cost=1743.09..1743.11 rows=1 width=32)
                  ->  Subquery Scan on clients_projects_1  (cost=1743.01..1743.09 rows=2 width=41)
                        ->  GroupAggregate  (cost=1743.01..1743.07 rows=2 width=81)
                              Group Key: projects_1.id
                              ->  Sort  (cost=1743.01..1743.01 rows=2 width=31)
                                    Sort Key: projects_1.id
                                    ->  Nested Loop Left Join  (cost=0.57..1743.00 rows=2 width=31)
                                          ->  Seq Scan on projects projects_1  (cost=0.00..1133.00 rows=2 width=17)
                                                Filter: (client_id = clients.id)
                                          ->  Nested Loop  (cost=0.57..5.00 rows=30000 width=18)
                                                ->  Index Only Scan using users_projects_pkey on users_projects  (cost=0.29..4.32 rows=2 width=8)
                                                      Index Cond: (project_id = projects_1.id)
                                                ->  Index Scan using users_pkey on users  (cost=0.29..0.34 rows=1 width=14)
                                                      Index Cond: (id = users_projects.user_id)
  
    $ # normal embed
    $ curl "localhost:3000/clients?select=*,projects(pi:id,pn:name,users(ui:id,un:name))" -H "Accept: application/vnd.pgrst.plan"
  
    Aggregate  (cost=35249539.07..35249539.09 rows=1 width=144)
      ->  Nested Loop Left Join  (cost=1174.94..35249389.07 rows=30000 width=48)
            ->  Seq Scan on clients  (cost=0.00..482.00 rows=30000 width=16)
            ->  Aggregate  (cost=1174.94..1174.95 rows=1 width=32)
                  ->  Nested Loop Left Join  (cost=20.94..1174.93 rows=2 width=49)
                        ->  Seq Scan on projects projects_1  (cost=0.00..1133.00 rows=2 width=17)
                              Filter: (client_id = clients.id)
                        ->  Aggregate  (cost=20.94..20.96 rows=1 width=32)
                              ->  Nested Loop  (cost=0.57..20.93 rows=2 width=14)
                                    ->  Index Only Scan using users_projects_pkey on users_projects  (cost=0.29..4.32 rows=2 width=4)
                                          Index Cond: (project_id = projects_1.id)
                                    ->  Index Scan using users_pkey on users  (cost=0.29..8.30 rows=1 width=14)
                                          Index Cond: (id = users_projects.user_id)
    ```
    
    </p>
    </details> 
  
    <details><summary>Responses</summary>
    <p>
    
    ```bash
    $ # to-many spread embed
    $ curl "localhost:3000/clients?select=*,projects(pi:id,pn:name,...users(ui:id,un:name))" -i
    
    HTTP/1.1 200 OK
    Transfer-Encoding: chunked
    Date: Wed, 09 Oct 2024 22:59:53 GMT
    Server: postgrest/12.3 (pre-release)
    Content-Range: 0-29999/*
    Content-Location: /clients?select=%2A%2Cprojects%28pi%3Aid%2Cpn%3Aname%2C...users%28ui%3Aid%2Cun%3Aname%29%29
    Content-Type: application/json; charset=utf-8
    Server-Timing: jwt;dur=0.0, parse;dur=0.7, plan;dur=0.3, transaction;dur=207.2, response;dur=0.0
  
    ...
  
    $ # normal embed
    $ curl "localhost:3000/clients?select=*,projects(pi:id,pn:name,users(ui:id,un:name))" -i
  
    HTTP/1.1 200 OK
    Transfer-Encoding: chunked
    Date: Wed, 09 Oct 2024 22:58:15 GMT
    Server: postgrest/12.3 (pre-release)
    Content-Range: 0-29999/*
    Content-Location: /clients?select=%2A%2Cprojects%28pi%3Aid%2Cpn%3Aname%2Cusers%28ui%3Aid%2Cun%3Aname%29%29
    Content-Type: application/json; charset=utf-8
    Server-Timing: jwt;dur=0.0, parse;dur=0.5, plan;dur=0.2, transaction;dur=216.7, response;dur=0.0
    
    ...
  
    ```
    
    </p>
    </details> 

    The to-many spread embed took 1:25~ minutes to respond. The normal embed also took 1:25~ minutes.

3. Now, adding one more nested relationship increases the cost by a ton, by more than 4000x the normal embed cost.

    <details><summary>Plan</summary>
    <p>
    
    ```console
    $ # to-many spread embed
    $ curl "localhost:3000/clients?select=*,projects(pi:id,pn:name,...users(ui:id,un:name,projects(pr:name)))" -H "Accept: application/vnd.pgrst.plan"

    Aggregate  (cost=465177233889.27..465177233889.29 rows=1 width=144)
      ->  Nested Loop Left Join  (cost=15505907.75..465177233739.26 rows=30000 width=48)
            ->  Seq Scan on clients  (cost=0.00..482.00 rows=30000 width=16)
            ->  Aggregate  (cost=15505907.75..15505907.77 rows=1 width=32)
                  ->  Subquery Scan on clients_projects_1  (cost=1.16..15505907.74 rows=2 width=41)
                        ->  GroupAggregate  (cost=1.16..15505907.72 rows=2 width=113)
                              Group Key: projects_1.id
                              ->  Merge Left Join  (cost=1.16..15505907.65 rows=2 width=63)
                                    Merge Cond: (projects_1.id = users_projects.project_id)
                                    ->  Index Scan using projects_pkey on projects projects_1  (cost=0.29..2100.29 rows=2 width=17)
                                          Filter: (client_id = clients.id)
                                    ->  Materialize  (cost=0.86..15503732.34 rows=30003 width=50)
                                          ->  Nested Loop Left Join  (cost=0.86..15503657.33 rows=30003 width=50)
                                                ->  Nested Loop  (cost=0.57..10925.20 rows=30003 width=18)
                                                      ->  Index Only Scan using users_projects_pkey on users_projects  (cost=0.29..786.33 rows=30003 width=8)
                                                      ->  Index Scan using users_pkey on users  (cost=0.29..0.34 rows=1 width=14)
                                                            Index Cond: (id = users_projects.user_id)
                                                ->  Subquery Scan on users_projects_3  (cost=0.29..516.36 rows=1 width=37)
                                                      ->  Nested Loop  (cost=0.29..516.35 rows=1 width=13)
                                                            ->  Seq Scan on users_projects users_projects_1  (cost=0.00..508.04 rows=1 width=4)
                                                                  Filter: (user_id = users.id)
                                                            ->  Index Scan using projects_pkey on projects  (cost=0.29..8.31 rows=1 width=17)
                                                                  Index Cond: (id = users_projects_1.project_id)
  
    $ # normal embed
    $ curl "localhost:3000/clients?select=*,projects(pi:id,pn:name,users(ui:id,un:name,projects(pr:name)))" -H "Accept: application/vnd.pgrst.plan"

    Aggregate  (cost=97214571.14..97214571.16 rows=1 width=144)
      ->  Nested Loop Left Join  (cost=3240.44..97214421.14 rows=30000 width=48)
            ->  Seq Scan on clients  (cost=0.00..482.00 rows=30000 width=16)
            ->  Aggregate  (cost=3240.44..3240.45 rows=1 width=32)
                  ->  Nested Loop Left Join  (cost=1053.69..3240.43 rows=2 width=49)
                        ->  Seq Scan on projects projects_1  (cost=0.00..1133.00 rows=2 width=17)
                              Filter: (client_id = clients.id)
                        ->  Aggregate  (cost=1053.69..1053.71 rows=1 width=32)
                              ->  Nested Loop Left Join  (cost=516.93..1053.68 rows=2 width=46)
                                    ->  Nested Loop  (cost=0.57..20.93 rows=2 width=14)
                                          ->  Index Only Scan using users_projects_pkey on users_projects  (cost=0.29..4.32 rows=2 width=4)
                                                Index Cond: (project_id = projects_1.id)
                                          ->  Index Scan using users_pkey on users  (cost=0.29..8.30 rows=1 width=14)
                                                Index Cond: (id = users_projects.user_id)
                                    ->  Aggregate  (cost=516.35..516.37 rows=1 width=32)
                                          ->  Nested Loop  (cost=0.29..516.35 rows=1 width=13)
                                                ->  Seq Scan on users_projects users_projects_1  (cost=0.00..508.04 rows=1 width=4)
                                                      Filter: (user_id = users.id)
                                                ->  Index Scan using projects_pkey on projects  (cost=0.29..8.31 rows=1 width=17)
                                                      Index Cond: (id = users_projects_1.project_id)
    ```
    
    </p>
    </details> 


    The time for the normal embed to respond is 2:00~ minutes. I stopped the to-many spread after 10:00~ minutes without completing.


So, as seen in point 3, there are ways to make the query not performant at all compared to the normal embed counterpart. I think a flag to activate the feature is still needed here.

Reply:
> The to-many spread took some milliseconds to respond. The normal embed took 1:30~ minutes.
> The to-many spread embed took 1:25~ minutes to respond. The normal embed also took 1:25~ minutes.

Hm, but these are already degenerate cases of API responses. I mean they wouldn't even have enough time to complete when postgREST is proxied by Cloudflare (default timeout is 60 seconds) and when PostgREST is hardened with a `statement_timeout` - which is a must and usually 10 seconds or less.

Reply:
> The to-many spread took some milliseconds to respond. The normal embed took 1:30~ minutes.
> The to-many spread embed took 1:25~ minutes to respond. The normal embed also took 1:25~ minutes.

Assuming a default `max-rows=1000`, how will these requests times look like?

Overall, I think we need to add statement_timeout and plan_filter to the hardening guide to avoid the above requests with such huge amount of rows. This is not specific to the to-many spread.

Reply:
> Assuming a default `max-rows=1000`, how will these requests times look like?

On `max-rows=1000` it improves a lot. The example 2 returns the data in 4~ seconds. The example 3 returns it in 18~ seconds.

> Hm, but these are already degenerate cases of API responses.

> Overall, I think we need to add statement_timeout and plan_filter to the hardening guide to avoid the above requests with such huge amount of rows. This is not specific to the to-many spread.

Hmm, yes, I agree.  There are other cases like deeply nesting normal embeds multiple times that are equally non-performant, e.g.  `/projects?select=users(*,projects(*,users(*,...)))`. So yeah, the problem is not specific to this feature, no need for the config here.